File: src/ciphers/FeistelCipher.java
package ciphers;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.Stack;

import data.Binary;

/*This is a sample implementation of the FeistelCipher.
 * The implementation assumes a block size of 32-bits.*/
public class FeistelCipher {
	String plainText;
	Stack<String> cipherTextBlocks;
	String key;
	String subKey;
	ArrayList<String> plainTextBlocks;

	public FeistelCipher(String plainText) {
		this.plainText = plainText;
		cipherTextBlocks = new Stack<String>();
		key = generateKey();
		subKey = key;
		plainTextBlocks = new ArrayList<String>();
		generateBlocks();
	}

	private String generateKey() {
		SecureRandom sr = new SecureRandom();
		String s = "";
		while (s.length() < 32) {
			s += sr.nextInt(2);
		}
		return s;
	}

	private String generateSubKey(boolean direction) {
		if (direction)
			subKey = Binary.circularShiftLeft(subKey);
		else
			subKey = Binary.circularShiftRight(subKey);
		return subKey;
	}

	private void generateBlocks() {
		String bytes = "" + plainText.charAt(0);
		for (int i = 1; i < plainText.length(); i++) {
			if (i % 4 == 0) {
				String binary = Binary.toBinary(bytes);
				plainTextBlocks.add(binary);
				bytes = "";
			}
			bytes += plainText.charAt(i);
		}
		plainTextBlocks.add(Binary.padWithZeros(Binary.toBinary(bytes), 32, false));
	}

	private String f(String rh, String subKey) {
		return Binary.xor(rh, subKey);
	}

	private String performRound(String lh, String rh, boolean encrypt) {
		String newrh = Binary.xor(f(rh, generateSubKey(encrypt)), lh);
		lh = rh;
		rh = newrh;
		return lh + "," + rh;
	}

	public void encrypt() {
		String cipherText = "";
		for (int i = 0; i < plainTextBlocks.size(); i++) {
			String block = plainTextBlocks.get(i);
			String lh = block.substring(0, 16);
			String rh = block.substring(16, 32);
			for (int j = 0; j < 16; j++) {
				String[] newBlock = performRound(lh, rh, true).split(",");
				lh = newBlock[0];
				rh = newBlock[1];
			}
			String temp = rh + lh;
			cipherTextBlocks.push(temp);
			cipherText += Binary.toBytes(temp);
		}
		generateSubKey(true);
		System.out.println("Encrypted Text: " + cipherText);
	}

	public void decrypt() {
		String originalText = "";
		while (!cipherTextBlocks.isEmpty()) {
			String block = cipherTextBlocks.pop();
			String lh = block.substring(0, 16);
			String rh = block.substring(16, 32);
			for (int i = 0; i < 16; i++) {
				String[] newBlock = performRound(lh, rh, false).split(",");
				lh = newBlock[0];
				rh = newBlock[1];
			}
			originalText = Binary.toBytes(rh + lh) + originalText;
		}
		System.out.println("Decrypted Text: " + originalText);
		System.out.println("Encryption made using a 32-bit block sized Feistel Cipher");
	}

	public static void main(String[] args) {
		FeistelCipher fc = new FeistelCipher("I love cookie <3");
		fc.encrypt();
		fc.decrypt();
	}
}

File: python/ciphers/FeistelCipher.py
import secrets
import data.Binary as Binary

class FeistelCipher:
    def __init__(self, plaintext):
        self.plaintext = plaintext
        self.cipher_text_blocks = []
        self.key = self.generate_key()
        self.subkey = self.key
        self.plaintext_blocks = []
        self.generate_plaintext_blocks()


    def generate_key(self):
        key = ""
        for i in range(0, 4):
            rand = str(secrets.randbelow(1000))
            key += Binary.to_binary(rand)
        return key[0: 32]

    def generate_subkey(self, direction):
        if direction:
            self.subkey = Binary.circular_shift_left(self.subkey)
        else:
            self.subkey = Binary.circular_shift_right(self.subkey)
        return self.subkey

    def generate_plaintext_blocks(self):
        word = ""
        for char in self.plaintext:
            if len(word) == 4:
                self.plaintext_blocks.append(Binary.to_binary(word))
                word = ""
            word += char
        self.plaintext_blocks.append(Binary.pad_zeros_left(Binary.to_binary(word), 32))

    def f(self, right_hand, subkey):
        return Binary.xor(right_hand, subkey)

    def perform_round(self, left_hand, right_hand, direction):
        new_right_hand = Binary.xor(left_hand, self.f(right_hand, self.generate_subkey(direction)))
        left_hand = right_hand
        right_hand = new_right_hand
        return left_hand, right_hand

    def encrypt(self):
        ciphertext = ""
        for plaintext_block in self.plaintext_blocks:
            left_hand = plaintext_block[0: 16]
            right_hand = plaintext_block[16:]
            for i in range(0, 16):
                new_block = self.perform_round(left_hand, right_hand, True)
                left_hand = new_block[0]
                right_hand = new_block[1]
            self.cipher_text_blocks.append(right_hand + left_hand)
            ciphertext += Binary.to_bytes(right_hand + left_hand)
        self.generate_subkey(True)
        print("Encrypted Text: " + ciphertext)

    def decrypt(self):
        plaintext = ""
        while self.cipher_text_blocks:
            ciphertext_block = self.cipher_text_blocks.pop()
            left_hand = ciphertext_block[0: 16]
            right_hand = ciphertext_block[16:]
            for i in range(0, 16):
                new_block = self.perform_round(left_hand, right_hand, False)
                left_hand = new_block[0]
                right_hand = new_block[1]
            plaintext = Binary.to_bytes(right_hand + left_hand) + plaintext
        print("Decrypted Text: " + plaintext)


f = FeistelCipher("Test1")
f.encrypt()
f.decrypt()
